src/desktop/space.rs

```diff
@@ -0,0 +1,419 @@
+    }
+
+    /// Iterate window in z-order back to front
+    pub fn windows(&self) -> impl Iterator<Item = &Window> {
+        self.windows.iter()
+    }
+
+    /// Get a reference to the window under a given point, if any
+    pub fn window_under(&self, point: Point<f64, Logical>) -> Option<&Window> {
+        self.windows.iter().find(|w| {
+            let loc = window_state(self.id, w).location;
+            let mut bbox = w.bbox();
+            bbox.loc += loc;
+            bbox.to_f64().contains(point)
+        })
+    }
+
+    pub fn window_for_surface(&self, surface: &wl_surface::WlSurface) -> Option<&Window> {
+        if !surface.as_ref().is_alive() {
+            return None;
+        }
+
+        self.windows
+            .iter()
+            .find(|w| w.toplevel().get_surface().map(|x| x == surface).unwrap_or(false))
+    }
+
+    pub fn window_geometry(&self, w: &Window) -> Option<Rectangle<i32, Logical>> {
+        if !self.windows.contains(w) {
+            return None;
+        }
+
+        Some(window_rect(w, &self.id))
+    }
+
+    pub fn map_output(&mut self, output: &Output, scale: f64, location: Point<i32, Logical>) {
+        let mut state = output_state(self.id, output);
+        *state = OutputState {
+            location,
+            render_scale: scale,
+            ..Default::default()
+        };
+        if !self.outputs.contains(output) {
+            self.outputs.push(output.clone());
+        }
+    }
+
+    pub fn outputs(&self) -> impl Iterator<Item = &Output> {
+        self.outputs.iter()
+    }
+
+    pub fn unmap_output(&mut self, output: &Output) {
+        if let Some(map) = output.user_data().get::<OutputUserdata>() {
+            map.borrow_mut().remove(&self.id);
+        }
+        self.outputs.retain(|o| o != output);
+    }
+
+    pub fn output_geometry(&self, o: &Output) -> Option<Rectangle<i32, Logical>> {
+        if !self.outputs.contains(o) {
+            return None;
+        }
+
+        let state = output_state(self.id, o);
+        o.current_mode().map(|mode| {
+            Rectangle::from_loc_and_size(
+                state.location,
+                mode.size.to_f64().to_logical(state.render_scale).to_i32_round(),
+            )
+        })
+    }
+
+    pub fn output_scale(&self, o: &Output) -> Option<f64> {
+        if !self.outputs.contains(o) {
+            return None;
+        }
+
+        let state = output_state(self.id, o);
+        Some(state.render_scale)
+    }
+
+    pub fn output_for_window(&self, w: &Window) -> Option<Output> {
+        if !self.windows.contains(w) {
+            return None;
+        }
+
+        let w_geo = self.window_geometry(w).unwrap();
+        for o in &self.outputs {
+            let o_geo = self.output_geometry(o).unwrap();
+            if w_geo.overlaps(o_geo) {
+                return Some(o.clone());
+            }
+        }
+
+        // TODO primary output
+        self.outputs.get(0).cloned()
+    }
```

Comment:
Shouldn't this really return a (possibly empty) set/vec of outputs, rather than just one output? Also, why does it return an output even if no output overlaps with the window?

Reply:
Yeah, that is a good idea. I think it always returns an `Output` if possible, because this method is used for fullscreen-ing and maximizing windows. But we could easily return a set/vec of outputs and just sort them by *most overlapping*? And also let downstream/anvil handle the fallback, if no `Output` overlaps.